src/mappings/event/tokens.ts

```diff
@@ -0,0 +1,461 @@
+
+    const fromAccount = address.interlay.encode(from);
+    const toAccount = address.interlay.encode(to);
+
+    entityBuffer.pushEntity(
+        Transfer.name,
+        new Transfer({
+            id: item.event.id,
+            height,
+            timestamp: new Date(block.timestamp),
+            from: fromAccount,
+            to: toAccount,
+            token: currency,
+            amount,
+            amountHuman,
+        })
+    );
+
+    const circulatingSupplyRangeFrom = getCirculatingSupplyProcessRange().range?.from || 0;
+    if (block.height < circulatingSupplyRangeFrom) {
+        // only start processing from the given height, otherwise exit early
+        return;
+    }
+
+    // if the transfer is in the native currency and to/from a system account, 
+    // we also want to update circulating supply counters
+    const nativeCurrency = getNativeCurrency();
+
+    if (!isSameCurrency(nativeCurrency, eventCcyId)) {
+        // not an event in the native currency (KINT/INTR); skip processing
+        return;
+    }
+
+    const isFromSysAccount = isSystemAddress(fromAccount);
+    const isToSysAccount = isSystemAddress(toAccount);
+    
+    if (isFromSysAccount === isToSysAccount) {
+        // poor man's !XOR:
+        // true === true: both to and from are system accounts, 
+        // false === false: neither is a system account
+        // in either case, nothing else to do because circulating supply doesn't change
+        return;
+    }
+        
+    const updateType = isFromSysAccount ? UpdateType.SystemSupplyDecrease : UpdateType.SystemSupplyIncrease;
+
+    const circulatingSupplyEntity = await updateCumulativeCirculatingSupply(
+        ctx,
+        block,
+        height,
+        nativeCurrency,
+        amount,
+        updateType,
+        entityBuffer
+    );
+
+    entityBuffer.pushEntity(CumulativeCirculatingSupply.name, circulatingSupplyEntity)
+}
+
+export async function tokensDeposited(
+    ctx: Ctx,
+    block: SubstrateBlock,
+    item: EventItem,
+    entityBuffer: EntityBuffer
+): Promise<void> {
+    const rawEvent = new TokensDepositedEvent(ctx, item.event);
+    let ccyId: CurrencyId_V17 | CurrencyId_V1020000 | CurrencyId_V1021000;
+    let accountId: Uint8Array;
+    let amount: bigint;
+
+    if (rawEvent.isV17) {
+        ({currencyId: ccyId, who: accountId, amount} = rawEvent.asV17);
+    } else if (rawEvent.isV1020000) {
+        ({currencyId: ccyId, who: accountId, amount} = rawEvent.asV1020000);
+    } else if (rawEvent.isV1021000) {
+        ({currencyId: ccyId, who: accountId, amount} = rawEvent.asV1021000);
+    } else {
+        ctx.log.warn(`UNKOWN EVENT VERSION: tokens.deposited`);
+        return;
+    }
+
+    const nativeCurrency = getNativeCurrency();
+
+    if (!isSameCurrency(nativeCurrency, ccyId)) {
+        // not an event in the native currency (KINT/INTR); skip processing
+        return;
+    }
+
+    const account = address.interlay.encode(accountId);
```

Comment:
Does this also encode correctly for kintsugi addresses?

Reply:
Yes, it's just an unfortunate historical name for the helper object:
```ts
export const address = {
    interlay: ss58.codec(ss58format),
    btc: {
        encode(address: AddressV6 | AddressV15): string | undefined {
            return encodeBtcAddress(address, bitcoinNetwork);
        },
    },
};
```

The codec and its prefix is set according to which network name is used in `.env` via subsquid's built-in registry.

Reply:
Refactored the name in commit c9a3943

`address.interlay` is now `address.parachain` to avoid confusion